### ISSUE TITLE
[discussion] log channel message structure

### ISSUE THREAD

**Issue by SimonWoolf** [ably/docs#309]
Example of what it is now:
```
Message {
  name: 'error',
  data: 
   { logData: 
      { type: 'ruleError',
        id: 'bv8Q1w|cbfL-6AwQABOVz59050111:0',
        ruleId: 'bv8Q1w',
        source: 'channel.message',
        channel: 'lambda_with_assumerole_cred_failure',
        messageId: 'fW2AShr1Lt:5',
        error: 'AccessDeniedException: User: arn:aws:iam::203461409171:user/realtime-ci-kinesis is not authorized to perform: lambda:InvokeFunction on resource: arn:aws:lambda:us-east-1:203461409171:function:wrong' },
     logMetadata: 
      { severity: 'error',
        tag: 'lambda',
        scopeId: 'channel:c9bzmg:lambda_with_assumerole_cred_failure',
        site: 'eu-west-1-a' } }
}
```
Was created with specifically reactor errors in mind, so may need to be tweaked if we want to use it for more general Ably issues. (e.g currently `error` is a string as it's generally an error from the kinesis/lambda libraries; for Ably issues we probably want an ErrorInfo with a readable `code`)

**Comment by mattheworiordan** [comment]
@SimonWoolf I don' really follow the thinking behind the separation of logData and logMetadata to be honest.  Given we will only publish log messages with different severities, I would go with something which is a superset of our `ErrorInfo` object:

```
Message {
  name: <Enum string> { 'fatal' | 'error' | 'warn' | 'info' | 'debug' },
  data: <ErrorInfo-like object> {
    severity: <Enum string>,
    code: <Int>,
    message: <String>,
    href: <String this always exists if code exists as it's a well known URL, see https://github.com/ably/wiki/issues/142>,
    statusCode: <Int? optional because that won't always be relevant>,
    cause: <ErrorInfo? for underlying ErrorInfo, see https://github.com/ably/docs/issues/242>,
    meta: <Arbitrary Key/Value string pairs with the following included by default> {
      tag: <String>,
      scopeId: <String>,
      site: <String>,
      ....
    }
  }
````

**Comment by SimonWoolf** [comment]
I don't have strong opinions on the structure, I just implemented @paddybyers 's proposal from  https://github.com/ably/realtime/pull/793#discussion_r92802074 . I believe the idea was that logData would be a variable object whose fields depend on the nature of the log message (one set of fields for reactor rules, with a ruleId, channel, messageId, and so on), wheras logMetadata would be fixed. (Yours is more structured, but I'm guessing Paddy will say it's a little too structured, with nowhere for cause-specific data like channelname and ruleId to go apart from stuffed them into the `message` string, making them hard to parse)

**Comment by paddybyers** [comment]
As discussed on the call:

```
Message {
  name: <Enum string> { 'fatal' | 'error' | 'warn' | 'info' | 'debug' },
  data: <object> {
    severity: <Enum string>,
    action: <String, represents message category>,
    message: <String, may contain occurrence-specific details>,
    error: <optional ErrorInfo> {
        message: <String>,
        code: <Int>,
        href: <String this always exists if code exists as it's a well known URL, see https://github.com/ably/wiki/issues/142>,
        statusCode: <Int? optional because that won't always be relevant>,
        cause: <ErrorInfo? for underlying ErrorInfo, see https://github.com/ably/docs/issues/242>,
    }
    meta: <Arbitrary Key/Value string pairs with the following included by default> {
      tag: <String>,
      scopeId: <String>,
      site: <String>,
      ....
    }
  }
````

**Comment by mattheworiordan** [comment]
Gets my vote 👍

**Comment by SimonWoolf** [comment]
Implementing this now, couple minor questions:

1. Are we sure we want to expose the scopeId in the metadata? AFAIK so far it's been purely internal, for customer visible things we've been using `appId:`, `accountId:`, and `channel:` (whichever is applicable). If we expose `scopeId` it means we'll no longer be as free to change it per our internal logging/filtering requirements, which we've been doing quite a lot (adding connection-wide scopes etc)

2. Do we need both 'action' and 'tag', and if so, what're the different purposes of each? (AFAICS they're both basically for categorising messages, so we may be better off with only one for simplicity). And do we have strong opinions on the format of it? Maybe something like `reactor.lambda.error`, `connection.inbound.rateLimitExceeded` `rest.tokenRequests.rateLimitExceeded`, etc.?

**Comment by mattheworiordan** [comment]
> Are we sure we want to expose the scopeId in the metadata? AFAIK so far it's been purely internal

Well I'm not sure to be honest.  If we don't think customers need it now, then let's rather omit it.  Less data we send the better.  Given meta attributes are meant to be arbitrary key/val pairs, we can always add easy when customers ned more info.

> Do we need both 'action' and 'tag'

I can't really comment on that right now as I don't know what is being used in realtime.  @paddybyers would have to advise on that.  I was only basing my structure on what I had seen before.

> Maybe something like `reactor.lambda.error`

Would that not just be `reactor.lambda` and the severity `error` though?  Whatever we decide, we really need to keep the list of tags in a public repo, perhaps in `ably-common` so that customers can reference the expected values.  WDYT?  Same goes for all attributes of these types I would say.  

> Do we really need two message strings (one standalone, one in the errorinfo)

@paddybyers will have to comment as I believe this change came about after chatting with him.  Saying that, I could easily see examples like:
 - outer: "WebHook delivery failed", inner: "Request timeout"
 - outer: "AMQP rule failed after 5 attempts", inner: "Authentication failure"

Given we should really do our best to keep a list of expected errors for customer by tag and place this somewhere (again docs or ably-common), I could see what we'd want to have two error messages.

**Comment by SimonWoolf** [comment]
> I can't really comment on that right now as I don't know what is being used in realtime.

it wouldn't be the same as in realtime -- `action` in realtime log messages is the class & method names, we don't want to expose those to customers

>I could easily see examples like:  outer: "WebHook delivery failed", inner: "Request timeout" outer: "AMQP rule failed after 5 attempts", inner: "Authentication failure"  Given we should really do our best to keep a list of expected errors  for customer by tag and place this somewhere (again docs or  ably-common), I could see what we'd want to have two error messages

In the proposed structure, the outer `message` has no `code`. So in your example, if we want people to use the log message to programmatically tell that a webhook delivery has failed (which is code 70001), we can't put that information in the outer message. (Which makes it not very useful, so why not get rid of it?)

ISTM that your examples are much better dealt with by using the `ErrorInfo.cause` mechanism, where you can have a nested ErrorInfo and both have codes; which the proposed log message also has.

**Comment by mattheworiordan** [comment]
> ISTM that your examples are much better dealt with by using the ErrorInfo.cause

@SimonWoolf sure, sounds fine to me.

**Comment by paddybyers** [comment]
> Are we sure we want to expose the scopeId in the metadata?

What does a customer do if he wants to be able to see log messages relating to a specific channel? Or connection? I realise that we will have to make those scope types stable if we expose it, but it is legitimate to expect metadata that allows messages to be grouped in useful ways.

> Do we need both 'action' and 'tag', and if so, what're the different purposes of each?

They've come about at different times so this isn't necessarily the historical rationale, but:

- tags are well-known - we can publish them, and customers can know that they are seeing messages relevant to certain functionality by looking for those tags;

- tags are broader and apply to a functional area, not an individual action; so under the `foo` tag you might have actions `foo.insert()`, `foo.update()`, `foo.delete()` etc.

> Do we really need two message strings (one standalone, one in the errorinfo)?

Log messages contain a message - that shouldn't be controversial. Not all log messages have an associated error. If they do happen to have an error, then why wouldn't we include the `ErrorInfo` in full, even if it also has a message?

**Comment by SimonWoolf** [comment]
> What does a customer do if he wants to be able to see log messages relating to a specific channel? Or connection?

Filter by `appId`, `channel`, or `connectionId`, as appropriate. I'm not suggesting we don't expose the information, just that we expose it in the way customers expect it to be exposed. For example, [amqpMessage includes `channel` and `appId` properties](https://github.com/ably/realtime/blob/2af40595b4d5be81b933885f95cede896cf758cc/common/lib/reactor/amqp/amqpmessage.js#L25-L26); surely it doesn't make sense to expose those two properties in a reactor message, but mix them together in a `scopeId` (which isn't a concept we've exposed to customers before) in a log message about that reactor message.

> Log messages contain a message - that shouldn't be controversial. Not all log messages have an associated error. If they do happen to have an error, then why wouldn't we include the ErrorInfo in full, even if it also has a message?

To be clear: you're suggesting here that when the log message is being used to wrap an ErrorInfo, the `message` be a copy of `error.message`? That'd be ok, just seems a bit redundant. But I guess better redundant than some generic "A reactor error has occurred, see the `error` for details", if it must be one of those two.

As noted previously, given the outer `message` has no `code` so isn't easy to use programatically, which may become a problem if we do want to start using this structure to deliver information which doesn't have an ErrorInfo `error`, but shrug.

**Comment by tbedford** [comment]
This doesn't look like a docs issue specifically, more a discussion around log message format. I don't see a specific documentation request here (so closing), but feel free to let me know otherwise.
